whisper/whisper05/api.go

```diff
@@ -0,0 +1,487 @@
+// Copyright 2016 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package whisper05
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/logger"
+	"github.com/ethereum/go-ethereum/logger/glog"
+	"github.com/ethereum/go-ethereum/rpc"
+
+	mathrand "math/rand"
+)
+
+// PublicWhisperAPI provides the whisper RPC service.
+type PublicWhisperAPI struct {
+	whisper *Whisper
+}
+
+// NewPublicWhisperAPI create a new RPC whisper service.
+func NewPublicWhisperAPI(w *Whisper) *PublicWhisperAPI {
+	return &PublicWhisperAPI{whisper: w}
+}
+
+// Version returns the Whisper version this node offers.
+func (self *PublicWhisperAPI) Version() (*rpc.HexNumber, error) {
+	if self.whisper == nil {
+		return rpc.NewHexNumber(0), whisperOffLineErr
+	}
+	return rpc.NewHexNumber(self.whisper.Version()), nil
+}
+
+// MarkPeerTrusted marks specific peer trusted, which will allow it
+// to send historic (expired) messages.
+func (self *PublicWhisperAPI) MarkPeerTrusted(peerID *rpc.HexBytes) error {
+	if self.whisper == nil {
+		return whisperOffLineErr
+	}
+	return self.whisper.MarkPeerTrusted(peerID)
+}
```

Comment:
I see that several APIs work with peer identifiers. We have not done this before because the peer set can change at any time and it can be hard to track that when interacting with the API via RPC. Can we find another way to do this?

Under which circumstances would an application call `shh_markPeerTrusted`? How does the application know that a certain peer is trustworthy?